java/arcs/android/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,210 @@
+            Operation.START_ARC -> runWithResult(intent) {
+                arcHost.startArc(
+                    intent.getParcelableExtra<ParcelablePlanPartition?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.STOP_ARC -> runWithResult(intent) {
+                arcHost.stopArc(
+                    intent.getParcelableExtra<ParcelablePlanPartition?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.REGISTER_PARTICLE -> runWithResult(intent) {
+                arcHost.registerParticle(
+                    intent.getParcelableExtra<ParcelableParticleIdentifier?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.UNREGISTER_PARTICLE -> runWithResult(intent) {
+                arcHost.unregisterParticle(
+                    intent.getParcelableExtra<ParcelableParticleIdentifier?>(
+                        OPERATION_ARG
+                    )?.actual!!
+                )
+            }
+            Operation.GET_REGISTERED_PARTICLES -> runWithResult(intent) {
+                arcHost.registeredParticles()
+            }
+            else -> {
+                throw IllegalArgumentException("Operation $intent not implemented.")
+            }
+        }
+    }
+
+    /**
+     * Execute an [ArcHost] method in the block, encode return into a [Bundle] and send via
+     * [ResultReceiver]
+     */
+    private suspend fun runWithResult(intent: Intent, block: suspend () -> Any?) {
+        val bundle = Bundle()
+        val result = block()
+
+        when (result) {
+            is List<*> -> {
+                val arrayList = ArrayList<ParcelableParticleIdentifier>()
+                result.forEach {
+                    it as ParticleIdentifier
+                    arrayList += it.toParcelable()
+                }
+                bundle.putParcelableArrayList(OPERATION_RESULT, arrayList)
+            }
+            is Parcelable -> bundle.putParcelable(OPERATION_RESULT, result)
+            is Serializable -> bundle.putSerializable(OPERATION_RESULT, result)
+            else -> {
+                // Unit
+            }
+        }
+        // This triggers a suspended coroutine to resume with the value.
+        intent.getParcelableExtra<ResultReceiver>(OPERATION_RECEIVER)?.send(0, bundle)
+    }
+
+    enum class Operation {
+        START_ARC, STOP_ARC, REGISTER_PARTICLE, UNREGISTER_PARTICLE,
+        GET_REGISTERED_PARTICLES
+    }
+
+    companion object {
+        const val ARC_HOST_INTENT = "arcs.android.host.ARC_HOST"
+        const val OPERATION = "OPERATION"
+        const val OPERATION_ARG = "OPERATION_ARG"
+        const val OPERATION_RECEIVER = "OPERATION_RECEIVER"
+        const val OPERATION_RESULT = "OPERATION_RESULT"
+
+        fun createArcHostIntent(
```

Comment:
These functions are where I think there's a good opportunity to use extension functions instead.

Reply:
Also, unless they're intended for public consumption/usage - consider restricting their visibility.

Reply:
They're used by AndroidManifestHostRegistry's inner class right now. I'll consider moving that inner class into ArcHostHelper, it's a private implementation detail.